nibabel/streamlines/trk.py

```diff
@@ -0,0 +1,814 @@
+    back on save.
+    """
+
+    # Contants
+    MAGIC_NUMBER = b"TRACK"
+    HEADER_SIZE = 1000
+    READ_BUFFER_SIZE = 87382 * 4  # About 4 Mb when no scalars nor properties.
+
+    def __init__(self, tractogram, header=None):
+        """
+        Parameters
+        ----------
+        tractogram : :class:`Tractogram` object
+            Tractogram that will be contained in this :class:`TrkFile`.
+
+        header : dict, optional
+            Metadata associated to this tractogram file.
+
+        Notes
+        -----
+        Streamlines of the tractogram are assumed to be in *RAS+*
+        and *mm* space where coordinate (0,0,0) refers to the center
+        of the voxel.
+        """
+        if header is None:
+            header_rec = TrkWriter.create_empty_header()
+            header = dict(zip(header_rec.dtype.names, header_rec[0]))
+
+        super(TrkFile, self).__init__(tractogram, header)
+
+    @classmethod
+    def get_magic_number(cls):
+        """ Return TRK's magic number. """
+        return cls.MAGIC_NUMBER
+
+    @classmethod
+    def support_data_per_point(cls):
+        """ Tells if this format supports saving data per point. """
+        return True
+
+    @classmethod
+    def support_data_per_streamline(cls):
+        """ Tells if this format supports saving data per streamline. """
+        return True
+
+    @classmethod
+    def is_correct_format(cls, fileobj):
+        """ Check if the file is in TRK format.
+
+        Parameters
+        ----------
+        fileobj : string or file-like object
+            If string, a filename; otherwise an open file-like object
+            pointing to TRK file (and ready to read from the beginning
+            of the TRK header data).
+
+        Returns
+        -------
+        is_correct_format : {True, False}
+            Returns True if `fileobj` is compatible with TRK format,
+            otherwise returns False.
+        """
+        with Opener(fileobj) as f:
+            magic_number = f.read(5)
+            f.seek(-5, os.SEEK_CUR)
+            return magic_number == cls.MAGIC_NUMBER
+
+        return False
+
+    @classmethod
+    def _create_arraysequence_from_generator(cls, gen):
+        """ Creates a ArraySequence object from a generator yielding tuples of
+            points, scalars and properties. """
+
+        streamlines = ArraySequence()
+        scalars = ArraySequence()
+        properties = np.array([])
+
+        gen = iter(gen)
+        try:
+            first_element = next(gen)
+            gen = itertools.chain([first_element], gen)
+        except StopIteration:
+            return streamlines, scalars, properties
+
+        # Allocated some buffer memory.
+        pts = np.asarray(first_element[0])
+        scals = np.asarray(first_element[1])
+        props = np.asarray(first_element[2])
+
+        scals_shape = scals.shape
+        props_shape = props.shape
+
+        streamlines._data = np.empty((cls.READ_BUFFER_SIZE, pts.shape[1]),
```

Comment:
This smells a bit fishy - this class method only uses `cls.READ_BUFFER_SIZE` from the class, and is also using private attributes from the ArraySequence class.

What is specific in this method, to the TrkFile class?   Can you think of a way to avoid using private methods of ArraySequence?